server.go

```diff
@@ -43,14 +57,91 @@ func NewServer(connBufferSize int, httpPort, grpcPort int) *Server {
 //Init starts the server
 func (s *Server) Init() {
 	go s.listen()
-	s.fetchBigBang()
+	if s.StateFileName != "" {
+		err := s.loadState(filepath.Join(statesFolderName, s.StateFileName))
+		if err != nil {
+			fmt.Println("\nLoading state failed, exiting now", err)
+			panic(err)
+		}
+	} else {
+		s.fetchBigBang()
+	}
 }
 
 //Run offloads the computation of changes to cis
 func (s *Server) Run() {
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+
 	for {
+		if len(signals) != 0 {
+			s.grpcServer.Stop()
+			fmt.Println("Received Signal:", <-signals)
+			break
+		}
 		s.step()
 	}
+
+	err := s.saveState(time.Now())
+	if err == nil {
+		fmt.Println("\nState successfully saved")
+	} else {
+		fmt.Println("\nState could not be saved:", err)
+	}
+}
+
+func createDirIfNotExist(dir string) error {
+	_, err := os.Stat(filepath.Join(dir))
+	if os.IsNotExist(err) {
+		err = os.MkdirAll(dir, 0755)
```

Comment:
> // MkdirAll creates a directory named path,
// along with any necessary parents, and returns nil,
// or else returns an error.
// The permission bits perm (before umask) are used for all
// directories that MkdirAll creates.
// If path is already a directory, MkdirAll does nothing
// and returns nil.

So you could just call `os.MkdirAll()` without checking if the dir already exists